### DISCUSSION TITLE
Parameter Store tags cannot be updated without losing the saved value

### DISCUSSION THREAD

**Discussion by adam-nielsen** [aws/aws-cdk#26033] [Q&A]
Summarising the discussion from https://github.com/aws/aws-cdk/issues/25949:

> If you change tags for a Parameter Store entry via CDK, the current value gets overwritten with the original value. This causes any passwords etc. that have been entered into Parameter Store entries to be lost.
>
> If you perform the same tag change via the AWS CLI or AWS Console, the value is left unchanged and no data is lost.
> ```
> const param = new ssm.StringParameter(this, `ssm`, {
> 	parameterName: `/example`,
> 	stringValue: 'TODO',
> 	tier: ssm.ParameterTier.STANDARD,
> });
> ```

(reproduction steps at the above issue link)

From @peterwoodworth:

> Yeah this is just how deploying CloudFormation templates works - they are deterministic, so if you specify your StringParameter to be this way, then it will be this way after deployment.

My issue is that the `stringValue` is mandatory, so you cannot omit it.  And it gets committed to git so you can't put the actual password in that you want to store.  This means any change you make to your stack risks wiping out your Parameter Store values and causing the deployed service to fail, e.g. because it can no longer retrieve the password it needs.

Perhaps this is a bug in CloudFormation?  Would it be better for it to be addressed by CloudFormation rather than by CDK?

**Comment by peterwoodworth** [comment]
`StringValue` is mandatory in [the SSM API](https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_PutParameter.html#systemsmanager-PutParameter-request-Value), so by extend CloudFormation and us require it as well when defining a new Parameter. 

> Perhaps this is a bug in CloudFormation

There isn't any bug here, CloudFormation is working as intended. Why are you defining this parameter in CloudFormation to begin with?

**Comment by adam-nielsen** [comment]
So let's say I have an application that needs to connect to a third party API.  Somehow I have to get the API credentials passed to the app.

I could hard-code these in the CDK stack and just pass them to Lambda or ECS as environment variables, which works fine, except then when I commit the CDK stack to git, I will end up with the actual credentials in the git repo which isn't great (for one, I can't then make the git repo public because it contains my private credentials).

So instead, in the stack I create two Parameter Store entries, one for the API username and one for the password.  I deploy the stack to our test, staging and production AWS accounts, which gives me two Parameter Store entries in each account.  I can then go in and manually set the appropriate API credentials (which are different for each account - the test account connects to the vendor's test API, production to their production API, etc.)  There are only two entries in this example but it's not uncommon for our code to have 10 or more entries for various APIs, external databases, etc. so this avoids having to manually create hundreds of Parameter Store entries across all the accounts for all our deployments.

This way the same code can be deployed to each account and it can retrieve the same Parameter Store entries by name, but each AWS account will get a different set of credentials and everything works perfectly.

Except when we later deploy an update to our stack, the Parameter Store entries will randomly get set back to the default value (in my case "TODO") instead of being left alone.

Is Parameter Store not designed for this use case?

I guess am a little confused why other things that store user data, like an S3 bucket or RDS database, do not get wiped and reset upon deploy, but Parameter Store does.  If Parameter Store values get reset when they 'drift', why not S3 buckets as well?

Alternatively, if Parameter Store is designed to always reflect the CDK stack and not allow manual updates, then what is its purpose?  If the value is meant to be hard-coded into the CDK stack, then why use Parameter Store at all over an environment variable passed to the code?  I thought the whole point of Parameter Store was to avoid having to hard-code credentials into the stack.

**Reply by peterwoodworth** [reply]
Thanks for the explanation @adam-nielsen, I see the problem you have and why you want to define this within CDK.

Since CloudFormation is declarative, creating the resources through CloudFormation will also make it so that they are managed through CloudFormation. This means that whenever you deploy your stack, the resources may get updated to accurately reflect their declared state.

You can alter the way you create your resources to reduce the likelihood they get modified after creation. One way you could do this is by refactoring the creation of these parameters into a Nested Stack, or at least an alternate stack, that only contains these parameters. This way, the stacks should be left alone after creation, but you still do run the risk that they get updated.

There's another way you can do this - I think you could instead make use of a custom resource. You could make your own custom resource which creates the parameters when the custom resource gets created, does nothing on update, and deletes them on deletion. Or, you could use [`AwsCustomResource`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.custom_resources.AwsCustomResource.html) since this is just one API call per parameter.

> Alternatively, if Parameter Store is designed to always reflect the CDK stack and not allow manual updates, then what is its purpose? If the value is meant to be hard-coded into the CDK stack, then why use Parameter Store at all over an environment variable passed to the code? I thought the whole point of Parameter Store was to avoid having to hard-code credentials into the stack.

I don't think the CloudFormation implementation is intended for creating parameters with sensitive information
